linkerd/stack/src/connect.rs

```diff
@@ -0,0 +1,124 @@
+use crate::{layer, Service};
+use futures::prelude::*;
+use linkerd_error::Error;
+use std::task::{Context, Poll};
+use tokio::io::{AsyncRead, AsyncWrite};
+
+/// A helper `Service` that drops metadata from a `MakeConnection`
+#[derive(Clone, Debug)]
+pub struct WithoutConnectionMetadata<S>(S);
+
+/// A helper that coerces a `MakeConnection` into a `Service`
+#[derive(Clone, Debug)]
+pub struct MakeConnectionService<S>(S);
+
+/// A helper trait that models a `Service` that creates client connections.
+///
+/// Implementers should implement `Service` and not `MakeConnection`. `MakeConnection` should only
+/// be used by consumers of these services.
+pub trait MakeConnection<T> {
+    /// An I/O type that represents a connection to the remote endpoint.
+    type Connection: AsyncRead + AsyncWrite;
+
+    /// Metadata associated with the established connection.
+    type Metadata;
+
+    type Error: Into<Error>;
+
+    type Future: std::future::Future<
```

Comment:
nit: i think `Future` is imported via `futures::prelude::*`?
```suggestion
    type Future: Future<
```